controller/src/main/java/org/jboss/as/controller/RestartParentResourceHandlerBase.java

```diff
@@ -57,19 +57,24 @@ public void execute(OperationContext context, ModelNode operation) throws Operat
                     final ServiceController<?> service = serviceName != null ?
                             context.getServiceRegistry(false).getService(serviceName) : null;
 
-                    ModelNode parentModel = null;
                     boolean servicesRestarted = false;
                     final boolean reloadRequired = service != null && !isResourceServiceRestartAllowed(context, service);
+                    ModelNode parentModel = reloadRequired || (service != null) ? getModel(context, address) : null;
                     if (reloadRequired) {
-                        parentModel = getModel(context, address);
                         if (parentModel != null) {
                             context.reloadRequired();
                         } // else the parent remove must have run as part of this op and we're not responsible for runtime
                     } else if (service != null ) {
-                        parentModel = getModel(context, address);
                         if (parentModel != null && context.markResourceRestarted(address, RestartParentResourceHandlerBase.this)) {
-                            removeServices(context, serviceName, parentModel);
-                            recreateParentService(context, address, parentModel);
+                            // Remove/recreate parent services in separate step using an OperationContext relative to the parent resource
+                            // This is necessary to support service installation via CapabilityServiceTarget
+                            context.addStep(Util.getReadResourceOperation(address), new OperationStepHandler() {
+                                @Override
+                                public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+                                    removeServices(context, serviceName, parentModel);
+                                    recreateParentService(context, parentModel);
+                                }
+                            }, OperationContext.Stage.RUNTIME);
```

Comment:
@pferraro I *think* these steps should be added at the top of the queue so they execute prior to any other steps, i.e. in the same relative position as the original code executed them. I pushed up #5528 to get CI going on that while we discuss.

Note that this commit is already merged into main, so please don't update it in this branch; if any change is appropriate it needs to be via a new commit.